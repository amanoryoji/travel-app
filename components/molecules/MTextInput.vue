<script lang="ts">
import { computed, PropType } from "vue";
export default {
  props: {
    /** 入力欄の種類 */
    type: {
      type: String as PropType<"text" | "number" | "password">,
      default: () => "text",
    },
    /** 入力欄にバインドする値 */
    modelValue: {
      type: [String, Number],
      default: () => "",
    },
    /** プレースホルダーの値 */
    placeholder: {
      type: String,
      default: () => "",
    },
    /** 入力パターン */
    pattern: {
      type: String,
    },
    /** 入力欄を幅いっぱい（width: 100%）にするか否か */
    block: {
      type: Boolean,
      default: () => false,
    },
    disabled: {
      type: Boolean,
      default: () => false,
    },
    /** 下部に表示するエラーメッセージ */
    errorMessage: {
      type: String,
    },
  },
  emits: ["update:modelValue", "focus", "blur", "change", "keydown"],
  setup(props, { emit }) {
    const model = computed({
      get: () => props.modelValue,
      set: (newValue) => {
        // 数値型なら数値に変更
        const value = props.type === "number" ? Number(newValue) : newValue;
        emit("update:modelValue", value);
      },
    });

    return {
      model,
    };
  },
};
</script>

<template>
  <div class="ly_textInputWrapper">
    <input
      :value="model"
      class="el_textInput"
      :type="type"
      :placeholder="placeholder"
      :pattern="pattern"
      :class="[
        { el_textInput__block: block },
        { el_textInput__error: !modelValue },
      ]"
      :disabled="disabled"
      @input="$emit('update:modelValue', $event.target.value)"
      @focus="$emit('focus')"
      @blur="$emit('blur')"
      @change="$emit('change', $event.target.value)"
      @keydown.enter="$emit('keydown')"
    />
    <AErrorMessage
      :message="errorMessage"
      font-size="sm"
      class="ly_errorMessage"
    />
  </div>
</template>

<style lang="scss" scoped>
.ly_textInputWrapper {
  display: flex;
  flex-direction: column;
}

.el_textInput {
  display: inline-block;
  height: 1.875rem;
  padding: 0.25rem;
  font-size: 1.25rem;
  border: none;
  border-bottom: 1px solid $color-input;

  &:focus {
    background-color: rgba($color-input, 0.2);
  }

  &::placeholder {
    color: $color-black-25;
  }

  // 幅いっぱいにする場合
  &__block {
    width: 100%;
    height: 100%;
  }

  &__error {
    border-bottom-color: $color-warning;
  }
}

.ly_errorMessage {
  margin-top: 8px;
  margin-left: 0.25rem;
}
</style>
