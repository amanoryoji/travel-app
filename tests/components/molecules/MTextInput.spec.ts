import { VueWrapper, shallowMount } from "@vue/test-utils";

import MTextInput from "/components/molecules/MTextInput.vue";


describe("BaseTextInput", () => {
  let wrapper: VueWrapper<MTextInput>;
  beforeEach(() => {
    wrapper = shallowMount(MTextInput);
  });

  it("コンポーネントがマウントされる", () => {
    expect(wrapper.vm).toBeTruthy();
  });

  describe("スナップショット", () => {
    it("デフォルト値", () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it("全propを指定", async () => {
      await wrapper.setProps({
        type: "text",
        value: "abc123",
        placeholder: "myPlaceholder",
        pattern: "^[0-9A-Za-z]+$",
        block: true,
        errorMessage: "errorMessage",
      });
      expect(wrapper.html()).toMatchSnapshot();
    });

    describe("type", () => {
      it("text", async () => {
        await wrapper.setProps({ type: "text" });
        expect(wrapper.html()).toMatchSnapshot();
      });

      it("number", async () => {
        await wrapper.setProps({ type: "number" });
        expect(wrapper.html()).toMatchSnapshot();
      });
    });
  });
});
