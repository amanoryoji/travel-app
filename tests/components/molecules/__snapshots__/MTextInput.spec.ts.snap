// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseTextInput > スナップショット > type > number 1`] = `
"<div data-v-16163742=\\"\\" class=\\"ly_textInputWrapper\\"><input data-v-16163742=\\"\\" class=\\"el_textInput el_textInput__error\\" type=\\"number\\" placeholder=\\"\\">
  <aerrormessage data-v-16163742=\\"\\" font-size=\\"sm\\" class=\\"ly_errorMessage\\"></aerrormessage>
</div>"
`;

exports[`BaseTextInput > スナップショット > type > text 1`] = `
"<div data-v-16163742=\\"\\" class=\\"ly_textInputWrapper\\"><input data-v-16163742=\\"\\" class=\\"el_textInput el_textInput__error\\" type=\\"text\\" placeholder=\\"\\">
  <aerrormessage data-v-16163742=\\"\\" font-size=\\"sm\\" class=\\"ly_errorMessage\\"></aerrormessage>
</div>"
`;

exports[`BaseTextInput > スナップショット > デフォルト値 1`] = `
"<div data-v-16163742=\\"\\" class=\\"ly_textInputWrapper\\"><input data-v-16163742=\\"\\" class=\\"el_textInput el_textInput__error\\" type=\\"text\\" placeholder=\\"\\">
  <aerrormessage data-v-16163742=\\"\\" font-size=\\"sm\\" class=\\"ly_errorMessage\\"></aerrormessage>
</div>"
`;

exports[`BaseTextInput > スナップショット > 全propを指定 1`] = `
"<div data-v-16163742=\\"\\" class=\\"ly_textInputWrapper\\" value=\\"abc123\\"><input data-v-16163742=\\"\\" class=\\"el_textInput el_textInput__block el_textInput__error\\" type=\\"text\\" placeholder=\\"myPlaceholder\\" pattern=\\"^[0-9A-Za-z]+$\\">
  <aerrormessage data-v-16163742=\\"\\" font-size=\\"sm\\" class=\\"ly_errorMessage\\" message=\\"errorMessage\\"></aerrormessage>
</div>"
`;
